File: py/control/gateway.py
from ..domain.contract import Contract
from ..api.proto import echoservice_pb2, echoservice_pb2_grpc

class Gateway(Contract):
    def __init__(self, channel):
        self.echo_stub = echoservice_pb2_grpc.EchoServiceStub(channel)

    def Echo(self, message: str) -> str:
        response = self.echo_stub.Echo(echoservice_pb2.EchoRequest(message=message))
        return response.message

File: go/pkg/control/gateway.go
package control

import (
	"context"

	"github.com/core-tools/hsu-echo/pkg/domain"
	"github.com/core-tools/hsu-echo/pkg/generated/api/proto"
	"github.com/core-tools/hsu-echo/pkg/logging"

	"google.golang.org/grpc"
)

func NewGRPCClientGateway(grpcClientConnection grpc.ClientConnInterface, logger logging.Logger) domain.Contract {
	grpcClient := proto.NewEchoServiceClient(grpcClientConnection)
	return &grpcClientGateway{
		grpcClient: grpcClient,
		logger:     logger,
	}
}

type grpcClientGateway struct {
	grpcClient proto.EchoServiceClient
	logger     logging.Logger
}

func (gw *grpcClientGateway) Echo(ctx context.Context, message string) (string, error) {
	response, err := gw.grpcClient.Echo(ctx, &proto.EchoRequest{Message: message})
	if err != nil {
		gw.logger.Errorf("Echo client gateway: %v", err)
		return "", err
	}
	gw.logger.Debugf("Echo client gateway done")
	return response.Message, nil
}
